zaptest/testingt.go

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zaptest
+
+// TestingT is a subset of the API provided by all *testing.T and *testing.B
+// objects.
+type TestingT interface {
```

Comment:
we'll be locked into this interface when we release it, so we might want to include methods that could be useful in future.

e.g., if we have an option to keep logs quiet unless the test fails, and we add a `Close` method that the user calls using `defer`, we'll want to check if the test has failed, so we might want to add the `Failed() bool` method now.

not sure if we need anything else (`Name() string` seems nice but don't have a concrete use case for it).

We could just take a `testing.TB` instead as well?

Reply:
Happy to take a `testing.TB` in the public API.

Reply:
This interface is a subset of `testing.TB`. We can't add custom methods
to it.

I'm happy to switch this to a different subset. FWIW, just `Errorf` and
`FailNow` have been enough in the past for testify. I added `Logf`
because we need it for this feature.

Depending directly on `testing.TB` here means that importing `zaptest`
always imports `testing`, which will always register extra flags on
your executable. I would like to avoid that if possible, but I don't
feel too strongly about it.

Reply:
I would also like to avoid importing the `testing` package due to the side effects.
If you are concerned about restricting zaptest to the smaller interface we could instead *copy* the entire `testing.TB` here rather than using a narrow subset. This should work nicely since the authors of the testing package are already "locked in" to this interface themselves and cannot change it easily as well.

The only minor downside I see is that such a wide interface with unused functions isn't exactly idiomatic.